x-pack/plugins/endpoint/server/services/endpoint/endpoint_query_builders.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { KibanaRequest } from 'kibana/server';
+import { EndpointAppContext } from '../../types';
+
+export class AllEndpointsQueryBuilder {
+  private readonly request: KibanaRequest<any, any, any>;
+  private readonly endpointAppContext: EndpointAppContext;
+
+  constructor(request: KibanaRequest<any, any, any>, endpointAppContext: EndpointAppContext) {
+    this.request = request;
+    this.endpointAppContext = endpointAppContext;
+  }
+
+  async toQueryParams(): Promise<Record<string, unknown>> {
+    const paging = await this.paging();
+    return {
+      body: {
+        query: this.queryBody(),
+        collapse: {
```

Comment:
why `collapse`? Can you leave a comment?

Reply:
> why collapse? Can you leave a comment?

We are technically aggregating by endpoint machine id and retrieving the latest of each group of events related to and endpoint by machine id. The collapse does this.

Reply:
Sounds good, can you just paste this in the source as well?